xtokens/src/lib.rs

```diff
@@ -395,16 +397,57 @@ pub mod module {
 			}
 		}
 	}
+
+	impl<T: Config> XcmTransfer<T::AccountId, T::Balance, T::CurrencyId> for Pallet<T> {
+		fn transfer(
+			who: T::AccountId,
+			currency_id: T::CurrencyId,
+			amount: T::Balance,
+			dest: MultiLocation,
+			dest_weight: Weight,
+		) -> XcmExecutionResult {
+			if amount.is_zero() {
+				return Ok(Outcome::Complete(Weight::zero()));
+			}
+			let id: MultiLocation =
+				T::CurrencyIdConvert::convert(currency_id).ok_or(Error::<T>::NotCrossChainTransferableCurrency)?;
+			let asset = MultiAsset::ConcreteFungible {
+				id,
+				amount: amount.into(),
+			};
+
+			with_xcm_execution_transaction(|| Self::do_transfer_multiasset(who, asset, dest, dest_weight))
+		}
+
+		fn transfer_multi_asset(
+			who: T::AccountId,
+			asset: MultiAsset,
+			dest: MultiLocation,
+			dest_weight: Weight,
+		) -> XcmExecutionResult {
+			if Self::is_zero_amount(&asset) {
+				return Ok(Outcome::Complete(Weight::zero()));
+			}
+
+			with_xcm_execution_transaction(|| Self::do_transfer_multiasset(who, asset, dest, dest_weight))
+		}
+	}
 }
 
-type XcmExecutionResult = sp_std::result::Result<Option<XcmError>, DispatchError>;
+type XcmExecutionResult = sp_std::result::Result<Outcome, DispatchError>;
 
 /// Only commit storage if no `DispatchError` and no `XcmError`, else roll back.
 fn with_xcm_execution_transaction(f: impl FnOnce() -> XcmExecutionResult) -> XcmExecutionResult {
 	with_transaction(|| {
 		let res = f();
 		match res {
-			Ok(ref err) if err.is_none() => TransactionOutcome::Commit(res),
+			Ok(ref outcome) => {
+				if let Outcome::Complete(_) = outcome {
+					TransactionOutcome::Commit(res)
+				} else {
+					TransactionOutcome::Rollback(res)
+				}
+			}
```

Comment:
```suggestion
			Ok(ref outcome) if outcome.ensure_complete().is_ok() => {
				TransactionOutcome::Commit(res)
			}
```

Reply:
`ensure_complete` takes ownership of `outcome`, so this suggestions would require a clone.

Reply:
Yeah you're right. In that case the outer `match` could be replaced with `if let` to get rid of the repeat `TransactionOutcome::Rollback(res)`, like:

```rust
if let Ok(ref outcome) = res {
    if let Outcome::Complete(_) = outcome {
        return TransactionOutcome::Commit(res);
    }
}
TransactionOutcome::Rollback(res)
```

Reply:
simplified even a bit more